### PR TITLE
refactor: Convert lib to TypeScript

### PR DESCRIPTION
**Pull request by sjinks**
## Description

Convert files from `src/lib` to TypeScript.

## Steps to Test

CI should pass.

### CODE REVIEW

File: src/bin/vip-import-sql.js [hunk 1]
```diff
@@ -573,7 +573,7 @@ Processing the SQL import for your environment...
 		} catch ( uploadError ) {
 			await track( 'import_sql_command_error', {
 				error_type: 'upload_failed',
-				upload_error: uploadError,
+				upload_error: uploadError.message,
```
Comment (sjinks) [inline src/bin/vip-import-sql.js:576]: We used to throw `string` from the code (OMG). I refactored the code to throw `Error` instead.

File: src/lib/api/http.ts [hunk 2]
```diff
@@ -14,6 +14,10 @@ import { API_HOST } from '../../lib/api';
 
 const debug = debugLib( '@automattic/vip:http' );
 
+type FetchOptions = Omit<RequestInit, 'body'> & {
+	body?: BodyInit | Record<string, unknown>;
+}
```
Comment (sjinks) [inline src/lib/api/http.ts:19]: Sometimes, we pass objects here. They will be encoded into strings, but we need a proper type definition because `RequestInit` does not like `body` as an `object`.

File: src/lib/search-and-replace.ts [hunk 3]
```diff
@@ -23,24 +22,18 @@ import { makeTempDir } from './utils';
 
 const debug = debugLib( '@automattic/vip:lib:search-and-replace' );
 
-const flatten = arr => {
-	return arr.reduce( ( flat, toFlatten ) => {
-		return flat.concat( Array.isArray( toFlatten ) ? flatten( toFlatten ) : toFlatten );
-	}, [] );
-};
```
Comment (sjinks) [inline src/lib/search-and-replace.ts:30]: No longer needed - we can use `.flatMap()` instead.

File: src/lib/search-and-replace.ts [hunk 4]
```diff
@@ -23,24 +22,18 @@ import { makeTempDir } from './utils';
...
+	fileName: string;
+	inPlace: boolean;
+	output: boolean | string | Writable;
+}
 
-export type GetReadAndWriteStreamsOutput = {
-	outputFileName?: string,
-	readStream: stream$Readable | Buffer,
-	usingStdOut: boolean,
-	writeStream: stream$Writable | Buffer,
```
Comment (sjinks) [inline src/lib/search-and-replace.ts:42]: There is no way these can be `Buffer` in the code.

File: src/lib/search-and-replace.ts [hunk 5]
```diff
@@ -130,7 +123,7 @@ export const searchAndReplace = async (
 	const fileSize = getFileSize( fileName );
 
 	// if we don't have any pairs to replace with, return the input file
-	if ( ! pairs || ! pairs.length ) {
+	if ( ! pairs.length ) {
```
Comment (sjinks) [inline src/lib/search-and-replace.ts:126]: `pairs: string | string[]`. Luckily, both `String` and `Array` have `.length` :-)

File: src/lib/client-file-uploader.ts [hunk 6]
```diff
@@ -53,51 +56,45 @@ export interface GetSignedUploadRequestDataArgs {
...
-export const getWorkingTempDir = async () =>
-	new Promise( ( resolve, reject ) => {
-		fs.mkdtemp( path.join( os.tmpdir(), 'vip-client-file-uploader' ), ( err, dir ) => {
-			if ( err ) {
-				return reject( err );
-			}
-			resolve( dir );
-		} );
-	} );
+const getWorkingTempDir = (): Promise<string> => mkdtemp( path.join( os.tmpdir(), 'vip-client-file-uploader' ) );
```
Comment (sjinks) [inline src/lib/client-file-uploader.ts:67]: Promise API rules!

File: src/lib/client-file-uploader.ts [hunk 7]
```diff
@@ -290,19 +319,15 @@ export async function uploadUsingMultipart( {
...
 	}
 
-	if (
-		! parsedResponse &&
-		parsedResponse.InitiateMultipartUploadResult &&
-		parsedResponse.InitiateMultipartUploadResult.UploadId
-	) {
-		throw `Unable to get Upload ID from cloud storage. Error: ${ multipartUploadResult }`;
+	if ( ! ('InitiateMultipartUploadResult' in parsedResponse) || ! parsedResponse.InitiateMultipartUploadResult.UploadId ) {
+		throw new Error( `Unable to get Upload ID from cloud storage. Error: ${ multipartUploadResult }` );
```
Comment (sjinks) [inline src/lib/client-file-uploader.ts:330]: The original condition was wrong: `parsedResponse` is always truthy.

File: src/lib/search-and-replace.ts [hunk 8]
```diff
@@ -170,34 +162,25 @@ export const searchAndReplace = async (
...
-
-				reject();
-			} );
-	} );
+	try {
+		await pipeline( replacedStream, writeStream );
+	} catch ( error ) {
+		console.log( red( "Oh no! We couldn't write to the output file.  Please check your available disk space and file/folder permissions." ) );
+		throw error;
+	}
```
Comment (sjinks) [inline src/lib/search-and-replace.ts:174]: I love the `pipeline()` function :-) It was introduced in Node 15; we require at least Node 16.

File: src/commands/dev-env-sync-sql.js [hunk 9]
```diff
@@ -174,8 +174,8 @@ export class DevEnvSyncSQLCommand {
 			await unzipFile( this.gzFile, this.sqlFile );
 			console.log( `${ chalk.green( '✓' ) } Extracted to ${ this.sqlFile }` );
 		} catch ( err ) {
-			await this.track( 'archive_extraction_error', { errorMessage: err?.message } );
```
Comment (pschoffer) [inline src/commands/dev-env-sync-sql.js:177]: Why this change? Is there a lint warning or something? We had a bug in upgrade wp site recently where library promised to never return null via types and it did.

**Review by pschoffer — CHANGES_REQUESTED** [review]

File: src/commands/dev-env-sync-sql.js (hunk 9 above)
Comment (sjinks) [reply src/commands/dev-env-sync-sql.js:177]: Because `err` will be an instance of `Error`, it cannot be `null` or `undefined`. See https://github.com/Automattic/vip-cli/pull/1391/files#diff-3e887d690f0e63a71d15440119be0462cb0d8b436c200a7425ea9f891d395dc7R106-R109

If `pipeline()` fails, it will reject with `Error`.

Comment (sjinks) [reply src/commands/dev-env-sync-sql.js:177]: > where library promised to never return null via types and it did

It is likely that type definitions for the library were compiled with [`strictNullChecks: false`](https://www.typescriptlang.org/tsconfig#strictNullChecks).

**Review by pschoffer — APPROVED** [review]
